File: week-5/course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        graph = [[] for i in range(numCourses)]
        visited = [False for i in range(numCourses)]
        
        for x, y in prerequisites:
            graph[x].append(y)
            
        for i in range(numCourses):
            if not self.dfs(i, graph, visited): #check for cycles in graph
                return False
            
        return True
    
    def dfs(self, i, graph, visited):
        if visited[i] == -1: # cycle found
            return False
        
        if visited[i]: return True # already visited node
        
        visited[i] = -1 #currently visiting node i
        
        for j in graph[i]:
            if not self.dfs(j, graph, visited):
                return False
            
        visited[i] = True # Finally mark node visited
        return True

File: week-5/course-schedule/course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    const graph = new Array(numCourses).fill(null).map(item=>[]);
    const visited = new Array(numCourses).fill(false);
    
    for (let [x, y] of prerequisites){
        graph[x].push(y);
    }

    for (let i=0; i<numCourses; i++) {
        if (!dfs(i, graph, visited)) return false;
    }
    
    return true;
};

const dfs = (i, graph, visited) => {
    if (visited[i] === -1) return false;
    
    if (visited[i]) return true;
    
    visited[i] = -1;
    for (let j of graph[i]) {
        if (!dfs(j, graph, visited)) return false;
    }
    
    visited[i] = true;
    return true;
};
